core/src/main/java/com/linecorp/armeria/common/stream/EventLoopStreamMessage.java

```diff
@@ -0,0 +1,402 @@
+        SubscriptionImpl subscription = this.subscription;
+        if (queue.isEmpty() && demand > 0 && !inOnNext) {
+            // Nothing in the queue and the subscriber is ready for an object, so send it directly.
+            demand--;
+            notifySubscriberOfObject(subscription, obj);
+            return;
+        }
+
+        doAddObjectOrEvent(obj);
+    }
+
+    private void doAddObjectOrEvent(Object obj) {
+        queue.add(obj);
+
+        if (subscription != null) {
+            doNotifySubscriber(subscription);
+        }
+    }
+
+    private void doNotifySubscriberIfNotEmpty() {
+        final SubscriptionImpl subscription = this.subscription;
+        if (subscription == null) {
+            return;
+        }
+
+        if (queue.isEmpty()) {
+            return;
+        }
+
+        doNotifySubscriber(subscription);
+    }
+
+    private void doNotifySubscriber(SubscriptionImpl subscription) {
+        if (inOnNext) {
+            // Do not let Subscriber.onNext() reenter, because it can lead to weird-looking event ordering
+            // for a Subscriber implemented like the following:
+            //
+            //   public void onNext(Object e) {
+            //       subscription.request(1);
+            //       ... Handle 'e' ...
+            //   }
+            //
+            // Note that we do not call this method again, because we are already in the notification loop
+            // and it will consume the element we've just added in addObject() from the queue as expected.
+            //
+            // We do not need to worry about synchronizing the access to 'inOnNext' because the subscriber
+            // methods must be on the same thread, or synchronized, according to Reactive Streams spec.
+            return;
+        }
+
+        if (!invokedOnSubscribe) {
+            // Subscriber.onSubscribe() was not invoked yet.
+            // Reschedule the notification so that onSubscribe() is invoked before other events.
+            eventLoop.execute(() -> doNotifySubscriber(subscription));
+            return;
+        }
+
+        for (;;) {
+            if (state == State.CLEANUP) {
+                cleanup();
+                return;
+            }
+
+            final Object o = queue.peek();
+            if (o == null) {
+                break;
+            }
+
+            if (o instanceof CloseEvent) {
+                doNotifySubscriberWithCloseEvent(subscription, (CloseEvent) o);
+                break;
+            }
+
+            if (demand == 0) {
+                break;
+            }
+
+            if (o instanceof AwaitDemandFuture) {
+                AwaitDemandFuture f = (AwaitDemandFuture) queue.remove();
+                f.complete(null);
+                continue;
+            }
+
+            demand--;
+
+            @SuppressWarnings("unchecked")
+            T obj = (T) queue.remove();
+            notifySubscriberOfObject(subscription, obj);
```

Comment:
`doNotifySubscriberOfObject()`? (no need to check if direct invocation is necessary?)

Reply:
I renamed the method since indeed it does need to be called in the event loop to set `inOnNext`. But either way, we do need to check direct invocation since the stream's event loop may be different from the subscriber's. Note that we are not executing any code outside the subscriber callbacks on the subscriber's event loop - it's all the stream's.